interfaces/connection.go

```diff
@@ -0,0 +1,222 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+type Connection struct {
+	plug *ConnectedPlug
+	slot *ConnectedSlot
+}
+
+type ConnectedPlug struct {
+	plugInfo     *snap.PlugInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type ConnectedSlot struct {
+	slotInfo     *snap.SlotInfo
+	dynamicAttrs map[string]interface{}
+}
+
+func NewConnectedSlot(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) (*ConnectedSlot, error) {
+	return &ConnectedSlot{
+		slotInfo:     slot,
+		dynamicAttrs: dynamicAttrs,
+	}, nil
+}
+
+func NewConnectedPlug(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) (*ConnectedPlug, error) {
+	return &ConnectedPlug{
+		plugInfo:     plug,
+		dynamicAttrs: dynamicAttrs,
+	}, nil
+}
+
+func (plug *ConnectedPlug) Interface() string {
+	return plug.plugInfo.Interface
+}
+
+func (plug *ConnectedPlug) Name() string {
+	return plug.plugInfo.Name
+}
+
+func (plug *ConnectedPlug) Snap() *snap.Info {
+	return plug.plugInfo.Snap
+}
+
+func (plug *ConnectedPlug) Apps() map[string]*snap.AppInfo {
+	return plug.plugInfo.Apps
+}
+
+func (plug *ConnectedPlug) Hooks() map[string]*snap.HookInfo {
+	return plug.plugInfo.Hooks
+}
+
+func (plug *ConnectedPlug) SecurityTags() []string {
+	return plug.plugInfo.SecurityTags()
+}
+
+func (plug *ConnectedPlug) StaticAttr(key string) (interface{}, error) {
+	if val, ok := plug.plugInfo.Attrs[key]; ok {
+		return copyRecursive(val)
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+func (plug *ConnectedPlug) StaticAttrs() (map[string]interface{}, error) {
+	return copyAttributes(plug.plugInfo.Attrs)
+}
+
+func (plug *ConnectedPlug) Attr(key string) (interface{}, error) {
+	if plug.dynamicAttrs != nil {
+		if val, ok := plug.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return plug.StaticAttr(key)
+}
+
+func (plug *ConnectedPlug) Attrs() (map[string]interface{}, error) {
+	if plug.dynamicAttrs != nil {
+		return plug.dynamicAttrs, nil
+	}
+	return nil, fmt.Errorf("dynamic attributes not initialized")
```

Comment:
It's not clear how this error can reasonably show up. If we have a ConnectedPlug at hand, we should have dynamic attributes because by definition we have a plug connected.

The message is not very user friendly, but it sounds like this should be made into an impossible situation instead.